test/core/client_channel/lb_policy/pick_first_test.cc

```diff
@@ -292,10 +308,12 @@ TEST_F(PickFirstTest, ShufflingDisabled) {
   constexpr static size_t kMaxAttempts = 5;
   for (size_t attempt = 0; attempt < kMaxAttempts; ++attempt) {
     absl::Status status = ApplyUpdate(
-        BuildUpdate(kAddresses, MakePickFirstConfig(true)), lb_policy_.get());
+        BuildUpdate(kAddresses, MakePickFirstConfig(false)), lb_policy_.get());
+    work_serializer_->Run([&]() { lb_policy_->ExitIdleLocked(); },
+                          DEBUG_LOCATION);
     EXPECT_TRUE(status.ok()) << status;
     std::vector<absl::string_view> address_order;
-    GetOrderAddressesArePicked(kAddresses, &address_order);
+    GetOrderAddressesArePicked(kAddresses, true, &address_order);
```

Comment:
We need to set `exit_idle_after_connection_fails` to false here.  It always needs to be set to false if we're going to do another pass after the current one.

If this is going to be too error prone, maybe we should just remove the `exit_idle_after_connection_fails` parameter and always use `ExpectState(GRPC_CHANNEL_IDLE)` instead of `ExpectStateAndQueueingPicker(GRPC_CHANNEL_IDLE)`.  That will skip checking the picker, but maybe that's okay in this case, since we already cover that case in the `GoesIdleWhenConnectionFailsThenCanReconnect` test.

Reply:
Done

Reply:
(tried multiple runs, seems perfectly stable)

Reply:
The parameter still needs to be set to false.

The problem here is not instability, it's that the test is not actually covering what it should.  The reason that you're not seeing any flakiness is that the first subchannel that the LB policy tries to connect when we exit idle is the same one that it will always try first after the next address list update.  But if the LB policy failed to try that particular address first after the update, the test could still pass, and in that case we would not want it to.

At this point, I think it's clear that we should remove the `exit_idle_after_connection_fails` parameter and always use `ExpectState(GRPC_CHANNEL_IDLE)` instead of `ExpectStateAndQueueingPicker(GRPC_CHANNEL_IDLE)`.  And we should add a comment indicating that we're explicitly *not* using the latter, because we don't want to trigger exiting idle until after the next address list update, if any.  Otherwise, the subtleties of the behavior here are going to be very unclear to future engineers who look at this code.

Reply:
Done.